pkg/beholder/client.go

```diff
@@ -0,0 +1,312 @@
+	retryDelay    time.Duration
+}
+
+type beholderClient struct {
+	config Config
+	// Logger
+	logger otellog.Logger
+	// Tracer
+	tracer oteltrace.Tracer
+	// Meter
+	meter otelmetric.Meter
+	// MessageEmitter
+	messageEmitter Emitter
+	// Graceful shutdown for tracer, meter, logger providers
+	closeFunc func() error
+}
+
+func NewClient(
+	config Config,
+	logger otellog.Logger,
+	tracer oteltrace.Tracer,
+	meter otelmetric.Meter,
+	messageEmitter Emitter,
+	onClose func() error,
+) Client {
+	return &beholderClient{
+		config:         config,
+		logger:         logger,
+		tracer:         tracer,
+		meter:          meter,
+		messageEmitter: messageEmitter,
+		closeFunc:      onClose,
+	}
+}
+
+// NewOtelClient creates a new BeholderClient with OTel exporter
+func NewOtelClient(cfg Config, errorHandler errorHandlerFunc) (Client, error) {
+	factory := func(ctx context.Context, options ...otlploggrpc.Option) (sdklog.Exporter, error) {
+		return otlploggrpc.New(ctx, options...)
+	}
+	return newOtelClient(cfg, errorHandler, factory)
+}
+
+// Used for testing to override the default exporter
+type otlploggrpcFactory func(ctx context.Context, options ...otlploggrpc.Option) (sdklog.Exporter, error)
+
+func newOtelClient(cfg Config, errorHandler errorHandlerFunc, otlploggrpcNew otlploggrpcFactory) (Client, error) {
+	ctx := context.Background()
+	baseResource, err := newOtelResource()
+	if err != nil {
+		return nil, err
+	}
+	sharedLogExporter, err := otlploggrpcNew(
+		ctx,
+		otlploggrpc.WithInsecure(),
+		otlploggrpc.WithEndpoint(cfg.OtelExporterGRPCEndpoint),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// Logger
+	loggerProcessor := sdklog.NewBatchProcessor(
+		sharedLogExporter,
+		sdklog.WithExportTimeout(1*time.Second), // Default is 30s
+	)
+	loggerAttributes := []attribute.KeyValue{
+		attribute.String("beholder_data_type", "zap_log_message"),
+	}
+	loggerResource, err := sdkresource.Merge(
+		sdkresource.NewSchemaless(loggerAttributes...),
+		baseResource,
+	)
+	if err != nil {
+		return nil, err
+	}
+	loggerProvider := sdklog.NewLoggerProvider(
+		sdklog.WithResource(loggerResource),
+		sdklog.WithProcessor(loggerProcessor),
+	)
+	logger := loggerProvider.Logger(cfg.PackageName)
+
+	// Set global logger provider
+	otelglobal.SetLoggerProvider(loggerProvider)
+
+	// Tracer
+	tracerProvider, err := newTracerProvider(cfg, baseResource)
+	if err != nil {
+		return nil, err
+	}
+	tracer := tracerProvider.Tracer(cfg.PackageName)
+	otel.SetTracerProvider(tracerProvider)
+	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(propagation.TraceContext{}, propagation.Baggage{}))
+
+	// Meter
+	meterProvider, err := newMeterProvider(cfg, baseResource)
+	if err != nil {
+		return nil, err
+	}
+	meter := meterProvider.Meter(cfg.PackageName)
+	otel.SetMeterProvider(meterProvider)
+
+	// MessageEmitter
+	messageLogProcessor := sdklog.NewBatchProcessor(
+		sharedLogExporter,
+		sdklog.WithExportTimeout(1*time.Second), // Default is 30s
```

Comment:
we should probably make this configurable as well (also applies to the meter and tracer providers in this file)

Reply:
There is a separate PR where this was addressed, will move it here.